packages/widgetbook/lib/src/addons/addon.dart

```diff
@@ -35,3 +37,54 @@ class WidgetbookAddOn {
   @override
   int get hashCode => name.hashCode;
 }
+
+abstract class WidgetbookAddOn2<T> {
+  WidgetbookAddOn2({
+    required this.name,
+    required this.setting,
+  });
+
+  final String name;
+  final T setting;
+  late ValueNotifier<T> provider;
+
+  T parseFromRouterData({
+    required Map<String, dynamic> routerData,
+    required T setting,
+  });
+
+  T get value => provider.value;
+
+  void onChanged(BuildContext context, T value) {
+    context.read<ValueNotifier<T>>().value = value;
+    context.goTo(queryParams: getQueryParameter(context));
+  }
+
+  Widget providerBuilder(
+    BuildContext context,
+    Map<String, dynamic> routerData,
+    Widget child,
+  ) {
+    final initialData = parseFromRouterData(
+      routerData: routerData,
+      setting: setting,
+    );
+    provider = ValueNotifier<T>(initialData);
+
+    return ChangeNotifierProvider.value(
+      key: ValueKey(initialData),
+      value: provider,
+      child: child,
+    );
+  }
+
+  Widget build(
+    BuildContext context,
+  );
+
+  Map<String, String> getQueryParameter(BuildContext context);
+}
+
+T readAddonValue<T>(BuildContext context) {
+  return context.watch<ValueNotifier<T>>().value;
+}
```

Comment:
🚨 This is the new implementation which a a major flaw. `readAddonValue` accesses the `provider` via the context. Due to the fact that I'm doing `context.watch<ValueNotifier<T>>()` we might fetch a wrong notifier here if we have to `WidgetbookAddOn2` of the same type.

```
class SoundAddOn extends WidgetbookAddOn2<bool> { ... }
class AdvertisementsAddOn extends WidgetbookAddOn2<bool> { ... }
```

@Roaa94 any ideas how we can improve this without making the API complex?

Reply:
Would it be too complex to have an abstract `AddonState` class that should be implemented by the state provided by the custom addon?

```dart
abstract class AddonState {
  const AddonState(this.value);

  final dynamic value;
}

abstract class WidgetbookAddOn2<T extends AddonState> {
   //...
}
```

Example implementation:
```dart
class SoundAddonState extends wb.AddonState {
  SoundAddonState(bool value) : super(value);
}

class SoundAddon extends wb.WidgetbookAddOn2<SoundAddonState> {
  SoundAddon()
      : super(
          name: 'Sound',
          setting: SoundAddonState(false),
        );

   //...
}
```